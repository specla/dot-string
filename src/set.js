/**
 * Set item value with dot-notation, this works for arrays and objects
 * @param {Array|Object} item
 * @param {String} path  [description]
 * @param {} value [description]
 */
export default function set (item, path, value) {
  if (typeof item !== 'object') {
    throw new TypeError('item should be the type of array or object')
  }

  if (typeof path !== 'string') {
    throw new TypeError('path should be a string')
  }

  path = path.split('.')
  const lastItem = path[path.length - 1]

  return path.reduce((acc, index) => {
    if (!acc[index] && index === lastItem) {
      acc[index] = value
    }

    if (!acc[index] && isNaN(index)) {
      acc[index] = {}
    }

    if (!acc[index] && !isNaN(index)) {
      acc[index] = []
    }

    return acc[index]
  }, item)
}
