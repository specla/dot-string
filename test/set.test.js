/* eslint-env jest */

import set from 'set'

test('Should set object property value', () => {
  const obj = { key: { item: undefined } }
  set(obj, 'key.item', true)
  expect(obj.key.item).toBe(true)
})

test('Should set array property value', () => {
  const array = [{ key: { item: undefined } }]
  set(array, '0.key.item', true)
  expect(array[0].key.item).toBe(true)
})

test('Should create path to key', () => {
  const array = []
  set(array, '0.key.item', true)
  expect(array[0].key.item).toBe(true)
})

test(`Should throw an error if the first argument isn't an array or object`, () => {
  expect(() => set('not-object', 'test')).toThrowErrorMatchingSnapshot()
})

test(`Should throw an error if the second argument isn't a string`, () => {
  expect(() => set({}, 623)).toThrowErrorMatchingSnapshot()
})
