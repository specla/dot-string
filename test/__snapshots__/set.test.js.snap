// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should throw an error if the first argument isn't an array or object 1`] = `"item should be the type of array or object"`;

exports[`Should throw an error if the second argument isn't a string 1`] = `"path should be a string"`;
